File: app/src/main/kotlin/tools/isekai/terminal/RelayCredentialVault.kt
package tools.isekai.terminal

import android.util.Base64

/**
 * relay_jwt(MASQUE relay経由P2P QUIC接続用のJWT、`ConnectionProfile.relayJwt`)を
 * Roomに保存する前に暗号化するための薄いラッパー。秘密鍵([KeyManager]/[KeystoreKek]参照)と
 * 同じ Android Keystore 由来の KEK(AES/GCM)を再利用する。
 *
 * 本格的な credential vault(`access_jwt` の短命化・メモリ限定保持、`refresh_token`/
 * `device_token` の発行・revoke/rotate、といった relay 認可サーバー前提の設計、
 * PLAN.md Phase 12以降の設計候補)が実装されるまでの、
 * [issue #1](https://github.com/cuzic/isekai-terminal/issues/1) に対する最小限の対策
 * (平文保存を無くすことだけを目的とする)。
 *
 * `AndroidKeyStore` は Robolectric では利用できないため、この object を直接呼ぶのは
 * 実際に Android フレームワークが動く経路(`AndroidAppExecutor`、`ProfileEditScreen` の
 * デフォルト引数)に限る。テストでは呼び出し元(`DumbAppExecutor` や
 * `ProfileEditScreen(encryptRelayJwt = { it }, decryptRelayJwt = { it })`)で
 * 恒等関数に差し替えること。
 */
object RelayCredentialVault {
    fun encrypt(plainJwt: String): String {
        KeystoreKek.generateKekIfAbsent()
        val enc = KeystoreKek.encrypt(plainJwt.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(enc, Base64.NO_WRAP)
    }

    fun decrypt(storedValue: String): String {
        val enc = Base64.decode(storedValue, Base64.NO_WRAP)
        return String(KeystoreKek.decrypt(enc), Charsets.UTF_8)
    }
}

File: ios/Sources/TsshCore/RelayCredentialVault.swift
import CryptoKit
import Foundation
import TsshCoreLogic

/// Phase 1E-6(#45): MASQUE relay経由P2P QUIC接続用のJWT(`ConnectionProfile.relayJwt`)を
/// GRDBに保存する前に暗号化するための薄いラッパー。Android版`RelayCredentialVault`+
/// `KeystoreKek`相当で、`CredentialVault`と同じKeychain由来のAES-GCM鍵機構
/// (`KeychainKEKStore`)を、秘密鍵ごとではなく固定の1鍵で再利用する。
///
/// 本格的なcredential vault(`access_jwt`の短命化・`refresh_token`/`device_token`の
/// 発行・revoke/rotate、relay認可サーバー前提の設計、PLAN.md Phase 12以降の設計候補)が
/// 実装されるまでの、平文保存を無くすことだけを目的とした最小限の対策
/// (Android版のissue #1対応コメント参照)。
public final class RelayCredentialVault {
    public enum VaultError: Error, Equatable {
        case keychainError(OSStatus)
        case decryptionFailed
    }

    private let store: KeychainKEKStore
    private let keyId = "relay-jwt-kek"

    public init(keychainService: String = "tools.isekai.terminal.relaycredentialvault") {
        self.store = KeychainKEKStore(service: keychainService)
    }

    public func encrypt(_ plainJwt: String) throws -> String {
        let kek = try store.getOrCreateKey(keyId: keyId)
        let sealedBox = try AES.GCM.seal(Data(plainJwt.utf8), using: kek)
        guard let combined = sealedBox.combined else { throw VaultError.decryptionFailed }
        return combined.base64EncodedString()
    }

    public func decrypt(_ storedValue: String) throws -> String {
        guard let combined = Data(base64Encoded: storedValue) else { throw VaultError.decryptionFailed }
        let kek = try store.getExistingKey(keyId: keyId)
        let sealedBox = try AES.GCM.SealedBox(combined: combined)
        let plaintext = try AES.GCM.open(sealedBox, using: kek)
        guard let jwt = String(data: plaintext, encoding: .utf8) else { throw VaultError.decryptionFailed }
        return jwt
    }
}
